domain/src/main/java/co/nimblehq/domain/usecase/base/BaseUseCase.kt

```diff
@@ -0,0 +1,38 @@
+package co.nimblehq.domain.usecase.base
+
+import co.nimblehq.domain.BuildConfig
+import co.nimblehq.domain.data.error.AppError
+import co.nimblehq.domain.data.error.NoConnectivityError
+import com.nimbl3.data.service.error.NoConnectivityException
+import okhttp3.internal.http2.StreamResetException
+import java.io.IOException
+
+abstract class BaseUseCase(
+    private val defaultError: (Throwable) -> AppError
+) {
+    protected fun composeError(error: Throwable): Throwable = when (error) {
+        is AppError -> error
+        is NoConnectivityException -> NoConnectivityError(error)
+        else -> defaultError(error)
+    }
+
+    protected fun doOnError(throwable: Throwable) {
+        if (BuildConfig.DEBUG) {
+            throwable.printStackTrace()
+        }
+    }
+
+    /** True if the request was canceled. */
+    protected fun isCanceledException(error: Throwable) =
+        isCanceledIOException(error) || isCanceledStreamResetException(error)
+
+    private fun isCanceledIOException(error: Throwable) =
+        error is IOException && error.message == MESSAGE_REQUEST_CANCELED
+
+    private fun isCanceledStreamResetException(error: Throwable) =
+        error is StreamResetException && error.message == MESSAGE_STREAM_WAS_RESET_CANCEL
+
+}
+
+private const val MESSAGE_REQUEST_CANCELED = "Canceled"
+private const val MESSAGE_STREAM_WAS_RESET_CANCEL = "stream was reset: CANCEL"
```

Comment:
Optional: Should this be place at the top of the class?

Reply:
@Tuubz I don't think so, like private methods in a class, we mostly put them on the bottom of the class. Those constants are not the things we would normally like to check when touching into a class for the first time and put them on top would bring annoying tbh 🤔

Reply:
Aha, my bad. Seems like I've been in the 7-Eleven project for too long 🙏 😁 I forgot that our guidelines specify to have it at the bottom

Reply:
in DeeMoney we put it on top as well 🤦🏻  @Tuubz

Reply:
@lydiasama the inspiration is from 7/11 I guess 😆